packages/hardhat/contracts/periphery/ControllerHelper.sol

```diff
@@ -295,178 +141,287 @@ contract ControllerHelper is UniswapControllerHelper, AaveControllerHelper, IERC
-                INonfungiblePositionManager.CollectParams({
-                    tokenId: _params.tokenId,
-                    recipient: address(this),
-                    amount0Max: type(uint128).max,
-                    amount1Max: type(uint128).max
-                })
-            )
-            : (wPowerPerpAmount, wethAmount) = INonfungiblePositionManager(nonfungiblePositionManager).collect(
-            INonfungiblePositionManager.CollectParams({
+                liquidity: _params.liquidity,
+                liquidityPercentage: 1e18,
+                amount0Min: uint128(_params.amount0Min),
+                amount1Min: uint128(_params.amount1Min)
+            }),
+            isWethToken0
+        );
+
+        if (wPowerPerpAmountInLp > 0) {
+            _exactInFlashSwap(
+                ControllerHelperDiamondStorage.getAddressAtSlot(4),
+                ControllerHelperDiamondStorage.getAddressAtSlot(5),
+                poolFee,
+                wPowerPerpAmountInLp,
+                _params.limitPriceEthPerPowerPerp.mul(wPowerPerpAmountInLp).div(1e18),
+                uint8(ControllerHelperDataType.CALLBACK_SOURCE.SWAP_EXACTIN_WPOWERPERP_ETH),
+                ""
+            );
+        }
+
+        ControllerHelperUtil.sendBack(
+            ControllerHelperDiamondStorage.getAddressAtSlot(5),
+            ControllerHelperDiamondStorage.getAddressAtSlot(4)
+        );
+    }
+
+    /**
+     * @notice Rebalance LP nft through trading
+     * @param _params ControllerHelperDataType.RebalanceWithoutVault struct
+     */
+    function rebalanceWithoutVault(ControllerHelperDataType.RebalanceWithoutVault calldata _params) external payable {
+        // if user need to send ETH to change LP composition, wrap to WETH
+        if (msg.value > 0) IWETH9(ControllerHelperDiamondStorage.getAddressAtSlot(5)).deposit{value: msg.value}();
+
+        INonfungiblePositionManager(ControllerHelperDiamondStorage.getAddressAtSlot(6)).safeTransferFrom(
+            msg.sender,
+            address(this),
+            _params.tokenId
+        );
+
+        // close LP NFT and get Weth and WPowerPerp amounts
+        (uint256 wPowerPerpAmountInLp, ) = ControllerHelperUtil.closeUniLp(
+            ControllerHelperDiamondStorage.getAddressAtSlot(6),
+            ControllerHelperDataType.closeUniLpParams({
                 tokenId: _params.tokenId,
-                recipient: address(this),
-                amount0Max: type(uint128).max,
-                amount1Max: type(uint128).max
-            })
+                liquidity: _params.liquidity,
+                liquidityPercentage: 1e18,
+                amount0Min: uint128(_params.amount0Min),
+                amount1Min: uint128(_params.amount1Min)
+            }),
+            isWethToken0
         );
 
-        if (wPowerPerpAmount > 0) {
+        ControllerHelperUtil.checkPartialLpClose(
```

Comment:
This is unnecessary if this function only is used for full closes and new ranges (as the liquidity percentage above is hardcoded as 1e18)

Is this function supposed to support re-using a NFT (same ranges, just more liquidity ie more ETH + oSQTH or not)?

Not sure this is to be added (but regardless should fix in different PR or new function, not this one :) ) [#301](https://github.com/opynfinance/squeeth-monorepo/issues/301)